src/ast.rs

```diff
@@ -0,0 +1,547 @@
+    pub expr: Option<AstNode<Expression>>,
+}
+
+/// A block.
+#[derive(Debug, Clone)]
+pub enum Block {
+    /// A match block.
+    Match(MatchBlock),
+    /// A loop block.
+    ///
+    /// The inner block is the loop body.
+    Loop(AstNode<Block>),
+    /// A body block.
+    Body(BodyBlock),
+}
+
+/// Function call arguments.
+#[derive(Debug, Clone)]
+pub struct FunctionCallArgs {
+    /// Each argument of the function call, as an expression.
+    pub entries: Vec<AstNode<Expression>>,
+}
+
+/// A function call expression.
+#[derive(Debug, Clone)]
+pub struct FunctionCallExpr {
+    /// An expression which evaluates to a function value.
+    pub subject: AstNode<Expression>,
+    /// Arguments to the function, in the form of [FunctionCallArgs].
+    pub args: AstNode<FunctionCallArgs>,
+}
+
+/// A logical operator.
+///
+/// These are treated differently from all other operators due to short-circuiting.
+#[derive(Debug, Clone, Eq, PartialEq, Copy)]
+pub enum LogicalOp {
+    /// The logical-and operator.
+    And,
+    /// The logical-or operator.
+    Or,
+}
+
+/// A logical operation expression.
+#[derive(Debug, Clone)]
+pub struct LogicalOpExpr {
+    /// The operator of the logical operation.
+    pub op: AstNode<LogicalOp>,
+    /// The left-hand side of the operation.
+    pub lhs: AstNode<Expression>,
+    /// The right-hand side of the operation.
+    pub rhs: AstNode<Expression>,
+}
+
+/// A property access exprssion.
+#[derive(Debug, Clone)]
+pub struct PropertyAccessExpr {
+    /// An expression which evaluates to a struct or tuple value.
+    pub subject: AstNode<Expression>,
+    /// The property of the subject to access.
+    pub property: AstNode<Name>,
+}
+
+/// A typed expression, e.g. `foo as int`.
+#[derive(Debug, Clone)]
+pub struct TypedExpr {
+    /// The annotated type of the expression.
+    pub ty: AstNode<Type>,
+    /// The expression being typed.
+    pub expr: AstNode<Expression>,
+}
+
+/// Represents a path to a module, given as a string literal to an `import` call.
+type ImportPath = String;
+
+/// A variable expression.
+#[derive(Debug, Clone)]
+pub struct VariableExpr {
+    /// The name of the variable.
+    pub name: AstNode<AccessName>,
+    /// Any type arguments of the variable. Only valid for traits.
+    pub type_args: Vec<AstNode<Type>>,
+}
+
+/// An expression.
+#[derive(Debug, Clone)]
+pub enum Expression {
+    /// A function call.
+    FunctionCall(FunctionCallExpr),
+    /// An intrinsic symbol.
+    Intrinsic(IntrinsicKey),
+    /// A logical operation.
+    LogicalOp(LogicalOpExpr),
+    /// A variable.
+    Variable(VariableExpr),
+    /// A property access.
+    PropertyAccess(PropertyAccessExpr),
+    /// A literal.
+    LiteralExpr(Literal),
+    /// A typed expression.
+    Typed(TypedExpr),
+    /// A block.
+    Block(AstNode<Block>),
+    /// An `import` call.
+    Import(AstNode<ImportPath>),
+}
+
+/// A module.
+///
+/// Represents a parsed `.hash` file.
+#[derive(Debug, Clone)]
```

Comment:
What does deriving `Debug` and `Clone` do to a particular struct?

Reply:
https://doc.rust-lang.org/rust-by-example/trait/derive.html